### ISSUE TITLE
Manual acknowledge of a single message

### ISSUE THREAD

**Issue by anjmao** [nats-io/nats-streaming-server#789]
I'm using QueueSubscribe with Manual Ack Mode. It works great and server redelivers message again in case it was not processed. But... Sometimes you know that there is a bug in you data and you can't consume the message. For this purpose I'm using a little program to ack a single message:

```go
var startAt uint64 = 1468
conn.Subscribe(subject, func(m *stan.Msg) {
    if m.Sequence == startAt {
        if err := m.Ack(); err != nil {
            log.Fatal(err)
        }
        fmt.Printf("ack: %s\n", string(m.Data))
    }
}, stan.StartAtSequence(startAt), stan.SetManualAckMode())
```

The problem is that it is not possible to specify to take only one message (limit) which is not ideal in case you have many new pending messages. What are recommended solutions to deal with such problem? Thanks.

**Comment by kozlovic** [comment]
You can somewhat limit the number of messages you want your subscriber to receive using `stan.MaxInflight(1)`. However, I am not sure you understand that messages do not get removed from the channel's message log just because they have been ack'ed. See https://github.com/nats-io/nats-streaming-server#message-log and following Subscriptions chapter.

In other words, suppose that you have a channel named `foo` with 10 messages (from 1 to 10). Then you start a subscription S1 that starts at 1, consume and ack all 10 messages. The subscription is still running (waiting for new messages). You then start a new subscription S2 that also starts at 1. It *will* receive all 10 messages because messages are in the log (and will be forever until channel limit force out old messages).
Ack'ing a message simply tells the server to stop redelivering this message to THIS subscription (or the queue group a member belongs to).

So, what you should do, is in your real application, if you detect that a message is not processable (miss some fields that you need, etc..) simply Ack() it:
```
  // Invoke this function that check that our message
  // can be successfully processed..
  if notAValidMsg(m) {
     if err := m.Ack(); err != nil {
       ...
     }
     // This is not a valid message, we are done.
     return
  }
  // Rest of the message processing logic
```

**Comment by anjmao** [comment]
@kozlovic Thanks for your answer. After posting my example I noticed the same that ACK will not work for simple subscribers. So if I understand correctly if I join the same QueueSubscriber group from my cmd util I should be able to ack the message, but only If STAN sends that message to my cmd and if I have many queue subscribers I could never get the message at all.

The problem with doing ack in real application is what I'm calling external services which is doing business logic validation so I can't really tell if fields are valid.

Ideally I would like STAN to create dead messages queue after some MAX retries so I could go and check if these messages are needed and data could be fixed by hand etc. I saw a PR for that but it was not accepted https://github.com/nats-io/gnatsd/pull/571 and sadly this is probably the only feature I miss in nats streaming.

**Comment by sujeshthekkepatt** [comment]
Any updates on this?

**Comment by kozlovic** [comment]
@sujeshthekkepatt In JetStream there will be worker queues and ability to NAck messages. I don't think there are plans for dead letter queue, but that could change.
In NATS Streaming, messages are not removed simply because they are ack'ed. As explained in earlier comment, ack'ing just means that the server will stop delivering those messages to a specific subscriber. You are free to create your own "error channel" and when a message cannot be processed, your app can send it to that channel and ack it in the subscription, this way, it won't be redelivered (again to *this* subscription). The bottom line is that a message somewhere in the message log cannot be removed (only the oldest messages are removed due to channel limits, not when ack'ed).

**Comment by kozlovic** [comment]
Closing this issue since in NATS Streaming, ack'ing a message never removes it from the channel, and this is by design.  Regarding multiple redelivery attempts, we have implemented a "redelivery count" (#996).
Closing this issue now.
